core/lib/contract_verifier/src/etherscan/client.rs

```diff
@@ -0,0 +1,329 @@
+use std::{collections::HashMap, time::Duration};
+
+use reqwest::Client;
+use serde::Serialize;
+use zksync_types::Address;
+
+use super::{
+    errors::{is_blocked_by_cloudflare_response, is_cloudflare_security_challenge, ApiError},
+    types::{ApiAction, ApiModule, ApiRequest, ApiResponseRaw, EtherscanVerificationRequest},
+};
+
+#[derive(Debug, Clone)]
+pub(crate) struct EtherscanClient {
+    api_url: String,
+    api_key: String,
+    http_client: Client,
+}
+
+fn process_api_call_result(
+    result: Result<impl AsRef<str>, reqwest::Error>,
+) -> Result<String, ApiError> {
+    match result {
+        Ok(result) => {
+            let result = result.as_ref();
+            let raw_response: ApiResponseRaw = serde_json::from_str(result).map_err(|error| {
+                if result.contains("Page not found") {
```

Comment:
I guess this can be checked using HTTP code (and in general we can handle everything that is not 200 here).

Reply:
I'm not sure Etherscan uses proper http code for _Page not found_ and other types of errors. In the refactored version I do check for 404 status code, but I also decided to leave this condition as well.